### ISSUE TITLE
[TF2.0] Loading a Saved Model failed with `AttributeError: '_UserObject' object has no attribute '_create_or_restore_slot_variable'`

### ISSUE THREAD

**Issue by hadim** [tensorflow/tensorflow#26814]
I am using the latest TF2.0 pip package.

Consider the following code:

```python
import tensorflow as tf
import numpy as np

class ToyModel(tf.keras.Model):
    """A simple linear model."""

    def __init__(self):
        super().__init__()
        self.l1 = tf.keras.layers.Dense(5)

    @tf.function(input_signature=[tf.TensorSpec([None, 1], tf.float32, name="input_func")])
    def call(self, x):
        return self.l1(x)

def toy_dataset():
    inputs = tf.range(10.)[:, None]
    outputs = inputs * 5. + tf.range(5.)[None, :]
    # TODO: switch `tuple` to `dict`.
    dataset = tf.data.Dataset.from_tensor_slices((inputs, outputs))
    return dataset.repeat(10).batch(2).shuffle(buffer_size=5)


dataset = toy_dataset()
optimizer = tf.keras.optimizers.Adam(0.1)

model = ToyModel()
model.compile(loss='mean_squared_error', optimizer=optimizer, metrics=['accuracy'])
history = model.fit(dataset, epochs=5)

# Inference works
inputs = np.array([[0, 5]], dtype=np.float32).T
print(model(inputs))

# Export to Saved Model
model_path = "/tmp/saved_model"
tf.saved_model.save(model, model_path)

# Load model
saved_model = tf.saved_model.load(model_path)

# Inference from Saved Model
#inputs = np.array([[0, 5]], dtype=np.float32).T
#saved_model(inputs)
```

It fails with:

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-4a1364881187> in <module>
     37 
     38 # Load model
---> 39 saved_model = tf.saved_model.load(model_path)
     40 
     41 # Inference from Saved Model

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/saved_model/load.py in load(export_dir, tags)
    322       loader = _Loader(object_graph_proto,
    323                        saved_model_proto,
--> 324                        export_dir)
    325       root = loader.get(0)
    326   else:

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/saved_model/load.py in __init__(self, object_graph_proto, saved_model_proto, export_dir)
     62     self._setup_functions_structures()
     63     self._setup_functions_captures()
---> 64     self._restore_checkpoint()
     65 
     66     for node in self._nodes:

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/saved_model/load.py in _restore_checkpoint(self)
    151     saver = util.TrackableSaver(graph_view.ObjectGraphView(self.get(0)))
    152     saver._file_prefix_placeholder = constant_op.constant(variables_path)
--> 153     load_status = saver.restore(variables_path)
    154     load_status.assert_existing_objects_matched()
    155     checkpoint = load_status._checkpoint

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/training/tracking/util.py in restore(self, save_path)
   1094         graph_view=self._graph_view)
   1095     base.CheckpointPosition(checkpoint=checkpoint, proto_id=0).restore(
-> 1096         self._graph_view.root)
   1097     load_status = CheckpointLoadStatus(
   1098         checkpoint,

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py in restore(self, trackable)
    207         # This object's correspondence with a checkpointed object is new, so
    208         # process deferred restorations for it and its dependencies.
--> 209         restore_ops = trackable._restore_from_checkpoint_position(self)  # pylint: disable=protected-access
    210         if restore_ops:
    211           self._checkpoint.new_restore_ops(restore_ops)

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py in _restore_from_checkpoint_position(self, checkpoint_position)
    777           ._single_restoration_from_checkpoint_position(
    778               checkpoint_position=current_position,
--> 779               visit_queue=visit_queue)))
    780     return restore_ops
    781 

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py in _single_restoration_from_checkpoint_position(self, checkpoint_position, visit_queue)
    804             child_position)
    805       else:
--> 806         if child_position.bind_object(trackable=local_object):
    807           # This object's correspondence is new, so dependencies need to be
    808           # visited. Delay doing it so that we get a breadth-first dependency

~/local/conda/envs/tf/lib/python3.7/site-packages/tensorflow/python/training/tracking/base.py in bind_object(self, trackable)
    249                       slot_name=slot_restoration.slot_name))
    250         else:
--> 251           optimizer_object._create_or_restore_slot_variable(  # pylint: disable=protected-access
    252               slot_variable_position=CheckpointPosition(
    253                   checkpoint=checkpoint,

AttributeError: '_UserObject' object has no attribute '_create_or_restore_slot_variable'
```

**Comment by hadim** [comment]
Maybe related to https://github.com/tensorflow/tensorflow/issues/25235 ?

**Comment by allenlavoie** [comment]
Sorry you ran into this; should be fixed by https://github.com/tensorflow/tensorflow/commit/147e6777068ea569cbecbd270b8b07c648bae09b

I checked with 2.0.0-dev20190318 and the save and load works. "call" isn't there with that name; we should make subclassed Models export any explicitly-decorated methods. [This method](https://github.com/tensorflow/tensorflow/blob/ac3faa26869258b6149f154618feb08dccd810ff/tensorflow/python/keras/engine/training.py#L1801) just needs to be modified to include [something like AutoTrackable's search for attributes containing decorated functions](https://github.com/tensorflow/tensorflow/blob/ac3faa26869258b6149f154618feb08dccd810ff/tensorflow/python/training/tracking/tracking.py#L94). We could have Layer inherit from AutoTrackable and call super().

**Comment by allenlavoie** [comment]
Or we could just add a copy of the traced function under "call" instead of just "_default_save_signature".

**Comment by hadim** [comment]
I confirm `2.0.0.dev20190318` fix the issue.

**Comment by hadim** [comment]
I let you close since you want to address something with `call`.

**Comment by allenlavoie** [comment]
Yep, leaving open to track the call issue. Thanks for confirming.

**Comment by allenlavoie** [comment]
Models export all their methods now ([example](https://github.com/tensorflow/tensorflow/blob/a37f3e3f88e1a91f9418ace165e59986164a942e/tensorflow/python/saved_model/load_test.py#L1554)). So well and truly fixed now.

**Comment by zenglian** [comment]
use `tf.keras.models.load_model(path)` instead.

**Comment by LatinAmericanProgramer** [comment]
I am having a similar issue. I recently installed anaconda 2020.02 on my Windows 10 laptop which comes with Spyder 4.0.1. .... then installed Tensorflow 2.1.0

`import tensorflow as tf
trained_model = tf.saved_model.load(export_dir='model_CNN_Dataset_pitch_tempo_envs1024xbatch_20xepochs_4')
trained_model.summary()
trained_model.layers`

it fails with:

> AttributeError: '_UserObject' object has no attribute 'summary'

Any suggestions please?

**Comment by allenlavoie** [comment]
@LatinAmericanProgramer if `summary` is from a Keras Model, I'd use the Keras save/load wrappers (`tf.keras.Model.save(..., save_format='tf')`, `tf.keras.load_model`). That will revive Models as Models rather than just an object with traced `tf.function`s. Otherwise pure Python methods are not saved with the SavedModel.

**Comment by LatinAmericanProgramer** [comment]
Thank you @allenlavoie ... I fix it uninstalling Tensorflow 2.1.0 and installing the version 2.2.0-dev20200327 using `pip install tf-nightly-gpu` . This fixed the problem for me.

Thank you

**Comment by bernardocortez** [comment]
I believe I have a similar problem... 

Python 3.7.6
TensorFlow 2.0.0

I am trying to save NN model ('Policy') with tf.keras.models.save_model and it raises AttributeError: 'Policy' object has no attribute 'built'.

Should I save it using another method?

Thanks!
